File: examples/spaced/ts/commands/telemetry.ts
/**
 * commands/telemetry.ts — telemetry get <mission>
 */

import { Command } from 'commander';
import { findMission, pick } from '../data';

interface TelemetryData {
  mission: string;
  vehicle: string;
  altitude_km: number;
  velocity_ms: number;
  downrange_km: number;
  stage: string;
  propellant_remaining_pct: number;
  comms_status: string;
  outcome: string;
  note: string;
}

function generateTelemetry(missionName: string, vehicle: string, outcome: string, note: string): TelemetryData {
  // Pseudo-random but deterministic per mission name
  const seed = missionName.split('').reduce((a, c) => a + c.charCodeAt(0), 0);
  const rand = (min: number, max: number) => min + (seed % (max - min + 1));

  const stages = ['STAGE_1_BURN', 'MECO', 'STAGE_SEP', 'STAGE_2_BURN', 'FAIRING_SEP', 'COAST'];

  let commsStatus: string;
  if (outcome === 'RUD') {
    commsStatus = 'LOSS_OF_SIGNAL (permanent)';
  } else if (outcome === 'PARTIAL') {
    commsStatus = 'DEGRADED';
  } else {
    commsStatus = 'NOMINAL';
  }

  return {
    mission: missionName,
    vehicle,
    altitude_km: rand(80, 550),
    velocity_ms: rand(4000, 7800),
    downrange_km: rand(200, 3000),
    stage: stages[seed % stages.length],
    propellant_remaining_pct: outcome === 'RUD' ? 0 : rand(12, 88),
    comms_status: commsStatus,
    outcome,
    note,
  };
}

function getFormat(cmd: Command): string {
  let c: Command | null = cmd;
  while (c) {
    const opts = c.opts();
    if (opts['format']) return opts['format'] as string;
    c = c.parent;
  }
  return 'table';
}

export function telemetryCommand(): Command {
  const telemetry = new Command('telemetry')
    .description('Mission telemetry streams');

  telemetry
    .command('get <mission>')
    .description('Get telemetry for a mission')
    .action(function (missionName: string) {
      const format = getFormat(this);
      const mission = findMission(missionName);

      if (!mission) {
        process.stderr.write(`  error: unknown mission: "${missionName}"\n`);
        process.stderr.write(`  Run 'spaced mission list' to see available missions.\n`);
        process.exit(1);
      }

      const data = generateTelemetry(mission.name, mission.vehicle, mission.outcome, pick(mission.notes));

      if (format === 'json') {
        console.log(JSON.stringify(data, null, 2));
        return;
      }

      if (format === 'yaml') {
        console.log(`mission: ${data.mission}`);
        console.log(`vehicle: ${data.vehicle}`);
        console.log(`altitude_km: ${data.altitude_km}`);
        console.log(`velocity_ms: ${data.velocity_ms}`);
        console.log(`downrange_km: ${data.downrange_km}`);
        console.log(`stage: ${data.stage}`);
        console.log(`propellant_remaining_pct: ${data.propellant_remaining_pct}`);
        console.log(`comms_status: ${data.comms_status}`);
        console.log(`outcome: ${data.outcome}`);
        console.log(`note: "${data.note}"`);
        return;
      }

      // Table
      console.log('');
      console.log(`  TELEMETRY: ${data.mission}`);
      console.log(`  ${'─'.repeat(50)}`);
      console.log(`  Vehicle:              ${data.vehicle}`);
      console.log(`  Altitude:             ${data.altitude_km} km`);
      console.log(`  Velocity:             ${data.velocity_ms} m/s`);
      console.log(`  Downrange:            ${data.downrange_km} km`);
      console.log(`  Stage:                ${data.stage}`);
      console.log(`  Propellant remaining: ${data.propellant_remaining_pct}%`);
      console.log(`  Comms:                ${data.comms_status}`);
      console.log(`  Outcome:              ${data.outcome}`);
      console.log('');
      console.log(`  Note: ${data.note}`);
      console.log('');
      if (data.outcome === 'RUD') {
        console.log('  ⚠ Telemetry stream terminated. Data archived for RCA.');
        console.log('    RCA timeline: 6–18 months. Updates via press conference.');
        console.log('');
      }
    });

  return telemetry;
}

File: examples/spaced/py/commands/telemetry.py
"""commands/telemetry.py — telemetry get <mission>."""
from __future__ import annotations

import json
import sys
from typing import Optional

import typer

sys.path.insert(0, __import__("os").path.join(__import__("os").path.dirname(__file__), "../../../../sdk/py"))
from data import find_mission  # noqa: E402

app = typer.Typer(help="Mission telemetry streams", add_completion=False)


def _mock_telemetry(mission_name: str) -> dict:
    import hashlib
    seed = int(hashlib.md5(mission_name.encode()).hexdigest(), 16) % 1000
    return {
        "mission": mission_name,
        "velocity_ms": 7823 + seed % 400,
        "altitude_km": 220 + seed % 180,
        "downrange_km": 1200 + seed % 2000,
        "propellant_pct": max(0, 88 - seed % 30),
        "engine_status": "nominal" if seed % 5 != 0 else "watching closely",
        "comms_signal": "strong" if seed % 4 != 0 else "intermittent",
        "g_force": round(1.2 + (seed % 20) / 10, 1),
        "source": "ground_network",
        "note": "telemetry is real-ish. the mission may not be.",
    }


@app.command("get")
def telemetry_get(
    mission: str = typer.Argument(..., help="Mission name"),
    format: Optional[str] = typer.Option(None, "--format", help="Output format: table|json|yaml"),
) -> None:
    """Retrieve live (simulated) telemetry for a mission."""
    m = find_mission(mission)
    if m is None:
        typer.echo(f"  ✗ Mission not found: {mission}", err=True)
        raise typer.Exit(1)

    telem = _mock_telemetry(m.name)
    fmt = (format or "table").lower()

    if fmt == "json":
        typer.echo(json.dumps(telem, indent=2))
        return

    if fmt == "yaml":
        import yaml  # noqa: PLC0415
        typer.echo(yaml.dump(telem, default_flow_style=False).rstrip())
        return

    # table
    typer.echo(f"""
  TELEMETRY: {m.name}  ({m.vehicle})
  {"─" * 50}
  Velocity:     {telem["velocity_ms"]} m/s
  Altitude:     {telem["altitude_km"]} km
  Downrange:    {telem["downrange_km"]} km
  Propellant:   {telem["propellant_pct"]}%
  Engines:      {telem["engine_status"]}
  Comms:        {telem["comms_signal"]}
  G-Force:      {telem["g_force"]}g
  Source:       {telem["source"]}

  Note: {telem["note"]}
""")
